torch/distributed/algorithms/fsdp_comm_hooks/allreduce_hook.py

```diff
@@ -0,0 +1,49 @@
+import torch
+import torch.distributed as dist
+
+
+class AllReduceState(object):
+    r"""
+    Stores parameters, needed to perform ``all_reduce`` algorithm
+    within a communication hook.
+    """
+
+    __slots__ = [
+        "process_group",
+        "world_size",
+        "gradient_predivide_factor",
+        "gradient_postdivide_factor"
+    ]
+
+    def __init__(
+        self,
+        process_group,
+        world_size
+    ):
+        self.process_group = process_group
+        self.world_size = world_size
+        self.gradient_predivide_factor = self._get_gradient_predivide_factor(
+            self.world_size
+        )
+        self.gradient_postdivide_factor = self.world_size / self.gradient_predivide_factor
+
+    # Same as in `FullyShardedDataParallel` class.
+    # Required to perform gradient pre-and post division
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
+            factor *= 2
```

Comment:
Let's copy the FSDP comment here:

```
    # setting two factors 'self.gradient_predivide_factor'
    # and 'self.gradient_postdivide_factor' to avoid underflow and overflow
```

@rohan-varma do you recall why this is sufficient to avoid overflow?

Reply:
Added comments. Discussed this with Rohan offline